dbms/src/Flash/ResourceControl/LocalAdmissionController.h

```diff
@@ -158,22 +336,220 @@ class ResourceGroup final : private boost::noncopyable
     // Total used cpu_time_in_ns of this ResourceGroup.
     uint64_t cpu_time_in_ns = 0;
 
+    TokenBucketMode bucket_mode = TokenBucketMode::normal_mode;
+
     double ru_consumption_delta = 0.0;
 
     LoggerPtr log;
+
+    std::chrono::time_point<std::chrono::steady_clock> last_fetch_tokens_from_gac_timepoint
+        = std::chrono::steady_clock::now();
+    std::chrono::time_point<std::chrono::steady_clock> stop_trickle_timepoint = std::chrono::steady_clock::now();
 };
 
 using ResourceGroupPtr = std::shared_ptr<ResourceGroup>;
 
+// LocalAdmissionController is the local(tiflash) part of the distributed token bucket algorithm.
+// It manages all resource groups:
+// 1. Creation, deletion and config updates of resource group.
+// 2. Fetching tokens from GAC periodically or when tokens are low.
+// 3. Record/report resource consumption and the priority of each resource group.
 class LocalAdmissionController final : private boost::noncopyable
 {
 public:
+    // For tidb_enable_resource_control is disabled.
+    static constexpr uint64_t HIGHEST_RESOURCE_GROUP_PRIORITY = 0;
+
+    LocalAdmissionController(::pingcap::kv::Cluster * cluster_, Etcd::ClientPtr etcd_client_)
+        : cluster(cluster_)
+        , etcd_client(etcd_client_)
+    {
+        background_threads.emplace_back([this] { this->startBackgroudJob(); });
+        background_threads.emplace_back([this] { this->watchGAC(); });
+    }
+
+    ~LocalAdmissionController() { stop(); }
+
+    void consumeResource(const std::string & name, double ru, uint64_t cpu_time_in_ns)
+    {
+        // When tidb_enable_resource_control is disabled, resource group name is empty.
+        if (name.empty())
+            return;
+
+        ResourceGroupPtr group = findResourceGroupWithException(name);
+        group->consumeResource(ru, cpu_time_in_ns);
+        if (group->lowToken())
+        {
+            {
+                std::lock_guard lock(mu);
+                low_token_resource_groups.insert(name);
+            }
+            cv.notify_one();
+        }
+    }
+
+    uint64_t getPriority(const std::string & name)
+    {
+        if (name.empty())
+            return HIGHEST_RESOURCE_GROUP_PRIORITY;
+
+        ResourceGroupPtr group = findResourceGroupWithException(name);
+        return group->getPriority(max_ru_per_sec.load());
+    }
```

Comment:
```suggestion
    std::optional<uint64_t> getPriority(const std::string & name)
    {
        if (name.empty())
            return HIGHEST_RESOURCE_GROUP_PRIORITY;

        ResourceGroupPtr group = findResourceGroup(name);
        if unlikely (group == nullptr)
            return {};
        return group->getPriority(max_ru_per_sec.load());
    }
```
I think it would be better to return optional, so that there is no need to catch exception in ResourceGroupQueue, but to do different processing according to the return value.
@windtalker What do you think about it?

Reply:
On the other hand, the cost of throwing exception is much more expensive than returning value.

Reply:
done